### PR TITLE
[Merged by Bors] - chore: forward port last part of leanprover-community/mathlib#18248

### PR DESCRIPTION
**Pull request by Felix-Weilacher**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

This is a forward port of a recent mathlib3 PR which added the following theorem: Any nonempty perfect set in a complete metric space admits an embedding of the Cantor space. 

I am very new to Lean 4, so I may need a lot of time or help.

* [`topology.perfect`@`4c19a16e4b705bf135cf9a80ac18fcc99c438514`..`49b7f94aab3a3bdca1f9f34c5d818afb253b3993`](https://leanprover-community.github.io/mathlib-port-status/file/topology/perfect?range=4c19a16e4b705bf135cf9a80ac18fcc99c438514..49b7f94aab3a3bdca1f9f34c5d818afb253b3993)

### CODE REVIEW

File: Mathlib/Topology/Perfect.lean [hunk 1]
```diff
@@ -214,3 +219,110 @@ theorem exists_perfect_nonempty_of_isClosed_of_not_countable [SecondCountableTop
...
+section CantorInj
+
+open Function ENNReal
+
+variable {α : Type _} [MetricSpace α] {C : Set α} (hC : Perfect C) {ε : ℝ≥0∞}
+
+private theorem Perfect.small_diam_aux (ε_pos : 0 < ε) {x : α} (xC : x ∈ C) :
+    let D := closure (EMetric.ball x (ε / 2) ∩ C)
+    Perfect D ∧ D.Nonempty ∧ D ⊆ C ∧ EMetric.diam D ≤ ε :=
+  by
```
Comment (semorrison) [inline Mathlib/Topology/Perfect.lean:234]:
```suggestion
    Perfect D ∧ D.Nonempty ∧ D ⊆ C ∧ EMetric.diam D ≤ ε := by
```

Comment (semorrison) [reply Mathlib/Topology/Perfect.lean:234]: And several times below.

File: Mathlib/Topology/Perfect.lean [hunk 2]
```diff
@@ -214,3 +219,110 @@ theorem exists_perfect_nonempty_of_isClosed_of_not_countable [SecondCountableTop
...
+    apply inter_subset_right
+  rw [EMetric.diam_closure]
+  apply le_trans (EMetric.diam_mono (inter_subset_left _ _))
+  convert EMetric.diam_ball (x := x)
+  rw [mul_comm, ENNReal.div_mul_cancel] <;> norm_num
+
+variable (hnonempty : C.Nonempty)
+
+/-- A refinement of `perfect.splitting` for metric spaces, where we also control
+the diameter of the new perfect sets. -/
```
Comment (semorrison) [inline Mathlib/Topology/Perfect.lean:252]:
```suggestion
/-- A refinement of `Perfect.splitting` for metric spaces, where we also control
the diameter of the new perfect sets. -/
```

File: Mathlib/Topology/Perfect.lean [hunk 3]
```diff
@@ -214,3 +219,110 @@ theorem exists_perfect_nonempty_of_isClosed_of_not_countable [SecondCountableTop
...
+  refine'
+    ⟨closure (EMetric.ball x₀ (ε / 2) ∩ D₀), closure (EMetric.ball x₁ (ε / 2) ∩ D₁),
+      ⟨perf0', non0', sub0'.trans sub0, diam0⟩, ⟨perf1', non1', sub1'.trans sub1, diam1⟩, _⟩
+  apply Disjoint.mono _ _ hdisj <;> assumption
+#align perfect.small_diam_splitting Perfect.small_diam_splitting
+
+open CantorScheme
+
+/-- Any nonempty perfect set in a complete metric space admits a continuous injection
+from the cantor space, `ℕ → bool`. -/
```
Comment (semorrison) [inline Mathlib/Topology/Perfect.lean:272]:
```suggestion
/-- Any nonempty perfect set in a complete metric space admits a continuous injection
from the cantor space, `ℕ → Bool`. -/
```

**Comment by semorrison** [comment]
It looks like there was an added comment in the mathlib3 version, that is missing here:

```
+* `set.scheme β α`: A `β`-scheme on `α`, a collection of subsets of `α` indexed by `list β`.
+  Used to construct maps `(β → ℕ) → α` as limiting objects.
```

File: Mathlib/Topology/Perfect.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Mathlib/Topology/Perfect.lean]: :+1:

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**